server/schedulers/evict_slow_trend.go

```diff
@@ -0,0 +1,437 @@
+			// slow node next time.
+			log.Info("store evicted by slow trend has been removed",
+				zap.Uint64("store-id", store.GetID()))
+			storeSlowTrendActionStatusGauge.WithLabelValues("evict.stop:removed").Inc()
+		} else if checkStoreCanRecover(cluster, store) {
+			log.Info("store evicted by slow trend has been recovered",
+				zap.Uint64("store-id", store.GetID()))
+			storeSlowTrendActionStatusGauge.WithLabelValues("evict.stop:recovered").Inc()
+		} else {
+			storeSlowTrendActionStatusGauge.WithLabelValues("evict.continue").Inc()
+			return s.scheduleEvictLeader(cluster), nil
+		}
+		s.cleanupEvictLeader(cluster)
+		return ops, nil
+	}
+
+	candFreshCaptured := false
+	if s.conf.candidate() == 0 {
+		candidate := chooseEvictCandidate(cluster)
+		if candidate != nil {
+			storeSlowTrendActionStatusGauge.WithLabelValues("cand.captured").Inc()
+			s.conf.captureCandidate(candidate.GetID())
+			candFreshCaptured := true
+		}
+	} else {
+		storeSlowTrendActionStatusGauge.WithLabelValues("cand.continue").Inc()
+	}
+	slowStoreID := s.conf.candidate()
+	if slowStoreID == 0 {
+		storeSlowTrendActionStatusGauge.WithLabelValues("cand.none").Inc()
+		return ops, nil
+	}
+
+	slowStore := cluster.GetStore(slowStoreID)
+	if !candFreshCaptured && checkStoreFasterThanOthers(cluster, slowStore) {
+		s.conf.popCandidate()
+		log.Info("slow store candidate by trend has been cancel",
+			zap.Uint64("store-id", slowStoreID))
+		storeSlowTrendActionStatusGauge.WithLabelValues("cand.cancel:too-faster").Inc()
+		return ops, nil
+	}
+	if !checkStoresAreUpdated(cluster, slowStore) {
+		log.Info("slow store candidate waiting for other stores to update heartbeats",
+			zap.Uint64("store-id", slowStoreID))
+		storeSlowTrendActionStatusGauge.WithLabelValues("cand.wait").Inc()
+		return ops, nil
+	}
+
+	candCapturedSecs := s.conf.candidateCapturedSecs()
+	log.Info("detected slow store by trend, start to evict leaders",
+		zap.Uint64("store-id", slowStoreID),
+		zap.Uint64("candidate-captured-secs", candCapturedSecs))
+	storeSlowTrendMiscGauge.WithLabelValues("cand.captured.secs").Set(float64(candCapturedSecs))
+	err := s.prepareEvictLeader(cluster, s.conf.popCandidate())
+	if err != nil {
+		log.Info("prepare for evicting leader by slow trend failed", zap.Error(err), zap.Uint64("store-id", slowStoreID))
+		storeSlowTrendActionStatusGauge.WithLabelValues("evict.prepare.err").Inc()
+		return ops, nil
+	}
+	storeSlowTrendActionStatusGauge.WithLabelValues("evict.start").Inc()
+	return s.scheduleEvictLeader(cluster), nil
+}
+
+func newEvictSlowTrendScheduler(opController *schedule.OperatorController, conf *evictSlowTrendSchedulerConfig) schedule.Scheduler {
+	return &evictSlowTrendScheduler{
+		BaseScheduler: NewBaseScheduler(opController),
+		conf:          conf,
+	}
+}
+
+func chooseEvictCandidate(cluster schedule.Cluster) (slowStore *core.StoreInfo) {
+	stores := cluster.GetStores()
+	if len(stores) <= 3 {
```

Comment:
One question here, why 3 nodes(normally, 3 is the common setting of TiKV cluster) is regarded as "too-few" nodes ?
IMO, `< 3` means too few nodes to make judgement whether there exists slow trends?

Reply:
Removed